### PR TITLE
[auto-materialize] Store and fetch historic rules

### PR DESCRIPTION
**Pull request by clairelin135**
Currently, the UI uses the current set of automaterialize rules for every past evaluation. This PR modifies the existing behavior by:
- serializing the existent rules as part of the asset evaluation
- loading the serialized rules in the UI, using the current set of rules when rules were not serialized in the evaluation for backcompat

### CODE REVIEW

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_auto_materialize_asset_evaluations.py [hunk 1]
```diff
@@ -154,16 +159,22 @@ def _test_get_evaluations(self, graphql_context: WorkspaceRequestContext):
             variables={"assetKey": {"path": ["asset_one"]}, "limit": 10, "cursor": None},
         )
         assert results.data == {
-            "autoMaterializeAssetEvaluationsOrError": {
```
Comment (clairelin135) [inline python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_auto_materialize_asset_evaluations.py:157]: This test is currently skipped on bk--refactor to allow this test to run

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 2]
```diff
@@ -625,6 +625,7 @@ class AutoMaterializeAssetEvaluation(NamedTuple):
     num_skipped: int
     num_discarded: int
     run_ids: Set[str] = set()
+    rule_snapshots: Optional[Sequence[AutoMaterializeRuleSnapshot]] = list()
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:628]: Allow this field to be optional to be able to deserialize old asset evaluations where this field was not populated

**Comment by clairelin135** [comment]
@OwenKephart here's a first stab at being able to display historic amp rules, to confirm that this is on the right track. No worries though if you don't get the chance to take a look before going on PTO

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:628]: nit: maybe default this to None? doesn't really make a difference but makes it easier to make sure the path of having it set to None works properly

**Review by OwenKephart — COMMENTED** [review]
overall, this looks great! this is probably already your intention, just logging it here that handling the case where there are no rule evaluation snapshots on the record can be done in the frontend (default to the current AMP ruleset)

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16123** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16123" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16123?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 2 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: good call

File: js_modules/dagster-ui/packages/ui-core/src/assets/AutoMaterializePolicyPage/RuleEvaluationOutcomes.tsx [hunk 3]
```diff
@@ -95,13 +95,13 @@ export const RuleEvaluationOutcomes = ({
           details={section.details}
         >
           <Box flex={{direction: 'column', gap: 8}}>
-            {(groupedRules[section.decisionType] || []).map(({description}) => {
```
Comment (clairelin135) [inline js_modules/dagster-ui/packages/ui-core/src/assets/AutoMaterializePolicyPage/RuleEvaluationOutcomes.tsx:98]: fixes a minor bug I encountered. descriptions may be the same for rules

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 4]
```diff
@@ -638,6 +639,11 @@ def from_rule_evaluation_results(
         num_discarded: int,
         dynamic_partitions_store: "DynamicPartitionsStore",
     ) -> "AutoMaterializeAssetEvaluation":
+        rule_snapshots = []
+        auto_materialize_policies = asset_graph.auto_materialize_policies_by_key.get(asset_key)
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:643]: this name should be singular, right?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 5]
```diff
@@ -638,6 +639,11 @@ def from_rule_evaluation_results(
...
+        if auto_materialize_policies:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:644]: will there never not be a policy? seems like we'd want to raise a check error in that case?

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/test_auto_materialize_asset_evaluation.py [hunk 6]
```diff
@@ -100,6 +100,11 @@ def test_backcompat():
         num_discarded=0,
         dynamic_partitions_store=None,
     )
+
+    # This field is populated via the auto materialize policies on the asset graph,
+    # replace to be None
+    expected_asset_evaluation.rule_snapshots._replace(rule_snapshots=None)
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/test_auto_materialize_asset_evaluation.py:106]: Does `_replace` mutate the object? I thought it just returns a new object and doesn't modify the underlying object. So this line wouldn't have any effect?

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_auto_materialize_asset_evaluations.py [hunk 7]
```diff
@@ -64,6 +64,10 @@
                         }
                     }
                 }
+                rules {
+                    decisionType
```
Comment (sryza) [inline python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_auto_materialize_asset_evaluations.py:68]: Might as well include class name in case we want to display it?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 8]
```diff
@@ -640,9 +640,10 @@ def from_rule_evaluation_results(
         dynamic_partitions_store: "DynamicPartitionsStore",
     ) -> "AutoMaterializeAssetEvaluation":
         rule_snapshots = []
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:642]: do we need this anymore?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 5 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: A lot of the old reconciliation scenario tests don't contain explicit auto materialize policies on the assets, and a check would cause a failure for all of those tests. 

I think for now we can rely on this function only being called for assets containing an auto-materialize policy, until we can migrate all of the tests to have auto materialize policies

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: Not the end of the world, but I don't love having code that's more complicated to support the tests.

How difficult would it be to update the tests? E.g. just have something in the test harness that adds an eager policy to all of the assets (and maybe a lazy policy instead for all of the assets in freshness_policy_scenarios.py).

Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: It's surprisingly tricky, I've spent a significant amount of time today to get these tests to pass but I think i'm almost there. Turns out there's a bunch of edge cases (i.e. some assets in a certain reconciliation scenario have AMPs, perf tests rely on snapshots, etc.)

Still need to look into additional test failures, i.e. this case is emitting an auto-materialize asset evaluation for asset1 even though it doesn't have a policy. Is it expected for an evaluation to exist here?: https://github.com/dagster-io/dagster/blob/652583c7995e059bb0936f01ab1abc1ce5c4cb26/python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/scenarios/auto_materialize_policy_scenarios.py#L58-L69

Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: Ah, looks like we auto-target all assets regardless of whether they have auto materialize policies, so it is expected that each asset has an evaluation

**Comment by clairelin135** [comment]
Closing this PR in favor of splitting into 2 parts:
- Adding implicit AMPs to tests: https://github.com/dagster-io/dagster/pull/16249
- Displaying historic AMP rules: https://github.com/dagster-io/dagster/pull/16250
